explanation/performance/perf-pgo.md

```diff
@@ -0,0 +1,252 @@
+
+## `perf` and AutoFDO
+
+Using `perf` to monitor a process and obtain data about its runtime workload produces data files in a specific binary format that we will call `perfdata`.  Unfortunately, GCC doesn't understand this file format; instead, it expects a profile file in a format called `gcov`.  To convert a `perfdata` file into a `gcov` one, we need to use a software called [`autofdo`](https://github.com/google/autofdo).  This software expects the binary being profiled to obey certain constraints:
+
+* The binary **cannot** be stripped of its debug symbols.  `autofdo` does not support separate debug information files (i.e., it can't work with Ubuntu's `.ddeb` packages), and virtually all Ubuntu packages run `strip` doing their build in order to generate the `.ddeb` packages.  If you intend to profile an Ubuntu package, please keep that in mind.
+
+* The debug information file(s) **cannot** be processed by `dwz`.  This is a tool whose purpose is to compress the debug information generated when building a binary, and again, virtually all Ubuntu packages use it.  For this reason, it is currently not possible to profile most of Ubuntu's packages without first rebuilding them to disable `dwz` from running.
+
+* We have to be mindful of the options we pass to `perf`, particularly when it comes to recording branch prediction events.  The options will likely vary depending on whether you are using an Intel or AMD processor, for example.
+
+On top of that, the current `autofdo` version in Ubuntu (`0.19-3build3`, at the time of this writing) is not recent enough to process the `perfdata` files we will generate.  There is a PPA with a newer version of `autofdo` package for Ubuntu Noble [here](https://launchpad.net/~sergiodj/+archive/ubuntu/autofdo).  If you are running another version of Ubuntu and want to install a newer version of `autofdo`, you will need to build the software manually (please refer to the [upstream repository](https://github.com/google/autofdo) for further instructions).
+
+## A simple PGO scenario: `openssl speed`
+
+PGO makes more sense when your software is CPU-bound, i.e., when it performs CPU intensive work and is not mostly waiting on I/O, for example.  Even if your software spends time waiting on I/O, using PGO might still be helpful; its effects would be less noticeable, though.
+
+OpenSSL has a built-in benchmark command called `openssl speed`, which tests the performance of its cryptographic algorithms.  This is excellent for PGO because there is practically no I/O involved, and we are just constrained by how fast the CPU can run.
```

Comment:
```suggestion
OpenSSL has a built-in benchmark command called `openssl speed`, which tests the performance of its cryptographic algorithms.  This is excellent for PGO because there is practically no I/O involved, and we are only constrained by how fast the CPU can run.
```

Reply:
("only" is less ambiguous than "just")